### PR TITLE
feat(wasm): let the loader ask the host for files (closes #2101)

### LINKED ISSUE #2101: [Architecture]: let the loader ask the host for files instead of the MCP server pre-collecting them
## Summary

The MCP server discovers a ledger's files itself — walking `include` directives, expanding globs, canonicalizing symlinks — and hands the result to `validateMultiFile` / `queryMultiFile` as a `{ path: contents }` map. #2100 moved include *semantics* to the loader, which fixed a family of bugs, but **discovery still lives in TypeScript** and that is where every remaining defect has been.

The fix is to let the loader ask the host for files instead, which is both the standard design for this problem and something rustledger is already 90% set up for.

## Why discovery keeps producing bugs

Every one of these was found in #2100, in the discovery layer, after the semantics had already been moved:

| symptom | cause |
|---|---|
| `ENOENT ... 'journals/*.beancount'` | globs not expanded |
| amounts doubled | a symlinked include counted twice |
| `file not found` on a whole ledger | symlinked **entry point** resolved against the link's directory |
| directory matched by `d/*` | `IsADirectoryError`-shaped failure |

None of these are beancount questions. They are filesystem questions that the loader already answers correctly for the CLI, re-answered by hand in another language.

## The standard design

For "the engine needs files, the host owns the filesystem", the established pattern is a host-implemented resolver rather than a pre-collected map:

- **Typst** — closest analogue, a Rust engine compiled to wasm with its own include system. The compiler never touches the filesystem; it asks a `World` trait for sources, and the CLI, the web/wasm build, the language server and the test harness each supply their own implementation.
- **Sass** — importer callbacks. **esbuild** — `onResolve` / `onLoad` plugins.
- **WASI** itself — no ambient authority; capabilities are granted explicitly by the embedder.

## What already exists

`rustledger_loader::FileSystem` is that trait, and its surface is already the right one:

```rust
fn read(&self, path: &Path) -> Result<Arc<str>, LoadError>;
fn exists(&self, path: &Path) -> bool;
fn dir_exists(&self, path: &Path) -> bool;
fn is_encrypted(&self, path: &Path) -> bool;
fn normalize(&self, path: &Path) -> PathBuf;
fn glob(&self, pattern: &str) -> Result<Vec<PathBuf>, String>;
fn decrypt(&self, path: &Path) -> Result<Arc<str>, LoadError>;
```

`DiskFileSystem` and `VirtualFileSystem` implement it, and #1667 already delegates decryption to a host capability, so host-provided behaviour across the boundary is not a new idea here.

Note `normalize`. That is exactly where symlink resolution belongs — and where the **macOS case-insensitivity gap** would be handled, which is currently unfixed because it cannot be reproduced or verified on Linux (see #2100). A host implementation running on the user's machine can answer it correctly without anyone guessing.

## Proposal

Add a wasm export that implements `FileSystem` over JS callbacks — `readFile`, `exists`, `glob`, `realpath` — supplied by the caller. `js_sys::Function` calls are synchronous, which matches the trait, so no async plumbing is required.

Then `collectLedgerFiles` and everything under it is **deleted**. The host provides primitives that encode no beancount semantics at all, and the loader does the walking.

## The constraint that kept this out of #2100

`packages/mcp-server` runs `npm ci` and builds against the **published** `@rustledger/wasm` (`^0.21.0`). A new export is unusable until it ships, whereas `validateMultiFile` / `queryMultiFile` already exist in 0.21.0 — which is why #2100 could take the semantics half immediately and had to leave this.

So this needs a wasm-crate change plus a release, in that order. It is not a redesign.

## Acceptance

With the host supplying only `readFile` / `exists` / `glob` / `realpath`, these must hold without any beancount-aware code in TypeScript:

- `include "journals/*.beancount"` expands
- a file reached via a symlink is loaded once, not twice
- a symlinked entry point resolves its includes against the real directory
- a diamond loads once and reports `Duplicate filename parsed`
- an error is attributed to the file and line it is actually on

All five are covered by tests in `packages/mcp-server/src/__tests__/index.test.ts` today, so the conversion has a ready-made check: they should keep passing while `collectLedgerFiles` disappears.

### PR DESCRIPTION
**Pull request by robcohen**
Closes #2101.

## What this adds

`validateWithHost(host, entry)` and `queryWithHost(host, entry, query)`. `host` is an object with three primitives, none of them beancount-aware:

- `readFile(path) => string | null` — required
- `glob(pattern) => string[]` — optional
- `realpath(path) => string` — optional, and the interesting one

The loader walks the include graph itself, asking for each file as it reaches it.

## Why

`validateMultiFile` takes every file up front, so the **caller** has to know which files a ledger contains — which means walking `include` directives itself. That is a second implementation of the loader's include handling, and the MCP server's copy was wrong four separate ways before #2100 replaced it: globs unexpanded, a file reached twice counted twice, a symlinked include counted twice, a symlinked entry resolved against the wrong directory.

Verified against `rledger check` on exactly the cases that each needed hand-written code before:

| | host path | `rledger` |
|---|---|---|
| glob include | 306.00 | 306.00 |
| diamond | 100.00 + `Duplicate filename parsed` | same |
| symlinked include | 10.00 + `Duplicate filename parsed` | same |
| symlinked entry | 10.00 | 10.00 |

**Nothing in this PR canonicalizes a path.** The host answers `realpath`, the loader de-duplicates by what it returns, and a file reached two ways loads once. That is the difference between a capability and a re-implementation — and it is also where the unfixed macOS case-insensitivity gap belongs, answerable by a host running on the user's machine rather than by me guessing on Linux.

## The `Send + Sync` problem, and how the repo already solved it

`FileSystem` requires `Send + Sync` for the rayon path, `js_sys::Function` is neither, and this crate is `#![forbid(unsafe_code)]`.

`HostFs` therefore holds **nothing** — a unit struct satisfies the bound for free — and the callbacks live in a thread-local installed for one call and cleared on drop, including on an early return. That is the same shape as `HostDecryptFs` in the WASI component (#1667): the capability is an ambient host function, not a handle carried in the struct.

Both entry points share the pipeline with their map-based counterparts rather than copying it. Giving either its own copy is precisely the mistake being removed.

## It also fixes a regression I introduced

#2100 made a duplicate include a `LoadError`, and `process` turns **every** load error into a blocking `LOAD` ledger error. So a wasm **query** over a ledger with a shared prices file included from two journals returned **no rows at all**, where `rledger query` prints the notice and answers. That is live on `main` and would have surfaced the moment the wasm was republished — for rustfava as well as the MCP server.

Only found by building the wasm and running it against real ledgers; CI tests the package against the *published* wasm, which predates the change.

`load_error_is_fatal` now gates the pre-processing abort, and the query path ignores `LOAD`-coded errors — matching the CLI in both directions: `check` still reports the duplicate and fails, `query` still returns the rows. `queryMultiFile` is fixed by the same change, not just the new entry points.

## On the tests

The first version of the duplicate-include test **passed with the regression restored** — reporting the duplicate and aborting on it produce the same symptoms, so asserting them proved nothing. The fixture now also fails a balance assertion: both errors present means the pipeline ran, the duplicate alone means it did not. Both tests were then verified to fail when `load_error_is_fatal` is reverted.

## Not included

Switching the MCP server over. It builds against the published `@rustledger/wasm`, so that is a follow-up after a release — the same ordering #2101 describes. `collectLedgerFiles` can be deleted then, and the five tests it added should keep passing unchanged.

### CODE REVIEW

**Comment by robcohen** [comment]
Review pass on the new code. Two real findings, both in `host_fs`.

## The `RefCell` borrow spanned every JS call

```rust
HOST.with(|h| {
    let borrowed = h.borrow();
    let value = Self::call1(&host.read_file, &display)?;  // JS runs HERE
```

A host callback is free to re-enter — a caching or logging wrapper is the obvious case — and re-entering while that borrow is live hits `borrow_mut` in `HostScope::install` and **panics**. In wasm that is a hard abort with a useless message.

Each site now clones the handle out and drops the borrow before calling. Cloning a `js_sys::Function` copies a handle, not a function.

## The `normalize` fallback reintroduced the doubling bug

When the host supplies no `realpath`, `normalize` only swapped backslashes. So `j/x.beancount` and `./j/../j/x.beancount` were **two** entries in the loader's already-loaded set — the file read twice, every amount in it doubled. That is the exact bug this whole design exists to prevent, reintroduced in the one path where the host cannot help.

It now applies the same clean-up the virtual backend does. Verified end to end with a host deliberately lacking `realpath`:

| | result | rledger check |
|---|---|---|
| with the fix | 10.00 | 10.00 |
| without | **20.00** | 10.00 |

Worth recording how nearly I missed that: my first attempt at the mutation emitted invalid Rust, the build failed, and the test ran against the **stale artifact** — reporting 10.00 and looking like a pass. I only caught it because `wasm-pack` printed `cargo build failed` above the result. Redone with code that compiles, which is where the 20.00 came from.

## Checked and clear

- **`dir_exists` returns `true`**, matching `VirtualFileSystem` — a flat map has no directory entries, and answering `false` would manufacture an E7006 on every load. The comment claiming that parity is accurate.
- **`load_error_is_fatal`** covers the variants that matter: duplicate-include survivable, cycle and path-traversal fatal.
- Unit tests for the normalizer cover `./`, `.`, `..`, backslashes and absolute paths.

File: crates/rustledger-wasm/src/host_fs.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        let realpath = js_fn(host, "realpath")?;
+
+        HOST.with(|h| {
+            *h.borrow_mut() = Some(HostCallbacks {
+                read_file,
+                glob,
+                realpath,
+            });
+        });
+        Ok(Self)
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-wasm/src/host_fs.rs:108]: Right, and I had fixed only half of this. I stopped a re-entrant call from panicking on the RefCell by releasing the borrow before calling into JS, but that left exactly the clobbering you describe — the inner scope overwrites, then clears on drop, and the outer load fails with 'no host filesystem installed' a long way from the cause. It now refuses the re-entrant install and says so. Verified: the message surfaces to the callback and the outer load still completes.

File: crates/rustledger-wasm/src/api.rs [hunk 2]
```diff
@@ -26,6 +26,23 @@ use crate::utils::LineLookup;
...
+///
+/// Treating every load error as fatal did exactly that — a query over a ledger
+/// with a shared prices file included from two journals returned NO rows,
+/// where the CLI returns them. The same shape as a warning being treated as an
+/// error, one layer down.
+fn load_error_is_fatal(error: &LoadError) -> bool {
+    !matches!(error, LoadError::DuplicateInclude { .. })
+}
+
 fn load_errors_to_errors(load_result: &LoadResult) -> Vec<Error> {
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-wasm/src/api.rs]: Correct — I inserted the new function between that doc comment and the body it belonged to, so it documented the wrong thing and left `load_errors_to_errors` undocumented. Moved back.

File: crates/rustledger-wasm/src/host_fs.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+/// A [`FileSystem`] that asks the JS host for each file as the loader reaches
+/// it. See the module docs for why it is a unit struct.
+#[derive(Debug)]
+pub struct HostFs;
+
+impl HostFs {
+    fn call1(f: &js_sys::Function, arg: &str) -> Result<JsValue, String> {
+        f.call1(&JsValue::NULL, &JsValue::from_str(arg))
+            .map_err(|e| {
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-wasm/src/host_fs.rs]: Good catch, and it would have broken a very ordinary host shape. The object the callbacks came from is now stored alongside them and used as the receiver. Verified with a host written as `{ readFile(p) { this.reads++; ... } }` — 306.00 and the counter incremented, where a null receiver would have thrown.

File: crates/rustledger-wasm/Cargo.toml [hunk 4]
```diff
@@ -62,6 +62,11 @@ getrandom.workspace = true
...
 console_error_panic_hook = "0.1"
 
+# Host-callback filesystem (#2101): reading `readFile`/`glob`/`realpath` off
+# the JS object the caller passes needs Reflect, Function and Array. Already a
+# dev-dependency for the wasm-bindgen tests; promoted so `host_fs` can use it.
+js-sys = "0.3"
+
 [dev-dependencies]
 wasm-bindgen-test = "0.3"
 js-sys = "0.3"
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-wasm/Cargo.toml]: Removed from dev-dependencies — it is a normal dependency now.
